File: core/src/vct/col/ast/BlockStatement.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import java.util.*;


public class BlockStatement extends ASTNode implements ASTSequence<BlockStatement> {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  private ArrayList<ASTNode> block=new ArrayList();
  
  public void add_statement(ASTNode s){
    add(s);
  }
  
  public int getLength(){ return block.size(); }
  
  public ASTNode getStatement(int i){ return block.get(i); }
  
  public ASTNode[] getStatements(){
    return block.toArray(new ASTNode[0]);
  }
  public boolean isEmpty(){
    return block.isEmpty();
  }

  @Override
  public Iterator iterator() {
    return block.iterator();
  }

  public void chain(ASTNode item){
    if (item instanceof BlockStatement){
      for(ASTNode n:((BlockStatement)item).block){
        n.clearParent();
        add(n);
      }
    } else {
      add(item);
    }
  }
  
  @Override
  public BlockStatement add(ASTNode item) {
    if (item!=null) {
      // this requires major rewriting elsewhere!
      //if (item instanceof ExpressionNode && !(item instanceof MethodInvokation)){
      //  hre.System.Failure("expressions must be wrapped in a Expression special to make them statements");
      //}
      block.add(item);
      item.setParent(this);
    }
    return this;
  }
  
  public BlockStatement append(ASTNode item) {
    if (item!=null) {
      if (item instanceof BlockStatement){
        for(ASTNode n:(BlockStatement)item){
          n.clearParent();
          this.add(n);
        }
      } else {
        this.add(item);
      }
    }
    return this;
  }

  @Override
  public int size() {
    return block.size();
  }

  @Override
  public ASTNode get(int i) {
    return block.get(i);
  }

  
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 
 

	public void prepend(ASTNode item) {
		block.add(0,item);
	  item.setParent(this);
  }

}

File: vercors/src/main/java/vct/col/ast/stmt/composite/BlockStatement.scala
package vct.col.ast.stmt.composite

import vct.col.ast.generic.{ASTNode, ASTSequence}
import vct.col.ast.util.ASTVisitor
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer

/**
 * AST node that represents a block of statements, that is, a sequence
 * "`S_1;...;S_n`" of (individual) statements `S_i`.
 */
class BlockStatement extends ASTNode with ASTSequence[BlockStatement] with VisitorHelper {
  /** The list of statements that constitutes the statement block. */
  private[this] val statements = new ArrayBuffer[ASTNode]
  
  /** 
   *  Adds a statement `stmt` to this block of statements.
   *  
   *  @return The resulting statement block (after adding `stmt`).
   */
  def addStatement(stmt:ASTNode) : BlockStatement = add(stmt)
  
  /** 
   *  Appends a given AST node "`node`", representing a statement, to this block of
   *  statements by applying chaining (see the method `chain`). 
   *  
   *  @return The resulting statement block (after appending `node`).
   */
  def append(node:ASTNode) : BlockStatement = chain(node)
  
  /** Identical to `get(i)`. */
  def getStatement(i:Int) = get(i)
  
  /** Yields a copy of the statements in this statement block. */
  def getStatements = statements.clone.toArray
  
  /** Tests whether the statement block is empty (does not contain any statements). */
  def isEmpty = statements.isEmpty
  
  /** Prepends the given AST node "`node`" to the front of this statement block. */
  def prepend(node:ASTNode) : Unit = prepend(Option(node))
  
  /** Applies the function "`f`" to each statement in this block (works the same as the
   *  `forEach` method provided by the [[Iterable]] interface, only `forEachStmt`
   * 	works nicer with Scala code). */
  def forEachStmt(f:ASTNode=>Unit) = statements.foreach(f)
  
  /** Identical to `size()`. */
  def getLength : Int = size
  
  /** If `element` contains an AST node, that node is added to 
   *  (the back of) this statement block (otherwise nothing is added). */
  def add(element:Option[ASTNode]) : Unit = element match {
    case Some(node) => node.setParent(this); statements += node
    case None => ()
  }
  
  /** 
   *  Adds the given AST node "`node`" to this block statement, and if `node` is
   *  a block statement itself, all `node`'s (internal) statements are added instead. 
   *  
   *  @param node The node to "chain" to the block statement
   *  @return The resulting statement block (after "chaining" `node`).
   */
  def chain(node:ASTNode) : BlockStatement = node match {
    case block:BlockStatement => block.forEachStmt { stmt => stmt.clearParent(); add(stmt) }; this
    case other => add(other)
  }
  
  /**
   * Prepends `element` to this statement block (i.e. `element` is added to 
   * the front of the statement list).
   * 
   * @note This method behaves differently from `append`, as `append` performs "chaining" 
   * (see the `chain` method) whereas `prepend` does not.
   */
  def prepend(element:Option[ASTNode]) : Unit = element match {
    case Some(node) => node.setParent(this); statements.+=:(node)
    case None => ()
  }

  /** 
   *  Adds "{@code node}" to (the back of) this statement block.
   *  
   *  @return The resulting statement block (after adding `node`).
   */
  override def add(node:ASTNode) : BlockStatement = { 
    add(Option(node))
    this 
  }
  
  /** Yields the `i`-th statement in this statement block. */
  override def get(i:Int) : ASTNode = statements.apply(i)
  
  /** Yields an (Java) iterator for the statements in this block. */
  override def iterator = statements.toIterator.asJava
  
  /** Yields the number of statements in this statement block. */
  override def size : Int = statements.length
  
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))

  override def debugTreeChildrenFields(): Iterable[String] = Seq("statements")
  override def debugTreePropertyFields(): Iterable[String] = Seq()
}
